File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/IndoorController.kt
package com.mapbox.maps.mapbox_maps

import com.mapbox.common.Cancelable
import com.mapbox.maps.IndoorFloor
import com.mapbox.maps.IndoorManager
import com.mapbox.maps.IndoorState
import com.mapbox.maps.MapboxExperimental
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.IndoorUpdatesStreamHandler
import com.mapbox.maps.mapbox_maps.pigeons.PigeonEventSink
import com.mapbox.maps.mapbox_maps.pigeons._IndoorMessenger
import io.flutter.plugin.common.BinaryMessenger

@OptIn(MapboxExperimental::class)
class IndoorController(
  private val mapboxMap: MapboxMap,
  messenger: BinaryMessenger,
  channelSuffix: String,
) : _IndoorMessenger {

  private var sink: PigeonEventSink<com.mapbox.maps.mapbox_maps.pigeons.IndoorState>? = null
  private var lastState =
    com.mapbox.maps.mapbox_maps.pigeons.IndoorState(floors = emptyList(), selectedFloorId = null)
  private var cancelable: Cancelable? = null

  init {
    val streamHandler = object : IndoorUpdatesStreamHandler() {
      override fun onListen(p0: Any?, sink: PigeonEventSink<com.mapbox.maps.mapbox_maps.pigeons.IndoorState>) {
        this@IndoorController.sink = sink
        sink.success(lastState)
      }

      override fun onCancel(p0: Any?) {
        sink = null
      }
    }
    IndoorUpdatesStreamHandler.register(messenger, streamHandler, channelSuffix)

    cancelable = mapboxMap.indoor.subscribeOnIndoorUpdated(object : IndoorManager.OnIndoorUpdatedCallback {
      override fun onIndoorUpdated(onIndoorUpdated: IndoorState) {
        lastState = onIndoorUpdated.toFLT()
        sink?.success(lastState)
      }
    })
  }

  override fun selectFloor(floorId: String?) {
    mapboxMap.indoor.selectFloor(floorId)
  }

  fun dispose() {
    cancelable?.cancel()
    cancelable = null
    sink = null
  }
}

private fun IndoorState.toFLT() = com.mapbox.maps.mapbox_maps.pigeons.IndoorState(
  floors = floors.map { it.toFLT() },
  selectedFloorId = selectedFloorId.takeUnless { it.isEmpty() },
)

private fun IndoorFloor.toFLT() = com.mapbox.maps.mapbox_maps.pigeons.IndoorFloor(id = id, name = name)

File: mapbox_maps_flutter_mobile/ios/mapbox_maps_flutter_mobile/Sources/mapbox_maps_flutter_mobile/Classes/IndoorController.swift
import Foundation
@_spi(Experimental) import MapboxMaps
import Flutter
import Combine

private final class IndoorEventStream: IndoorUpdatesStreamHandler {
    var sink: PigeonEventSink<IndoorState>?
    var lastState = IndoorState(floors: [])

    override func onListen(withArguments _: Any?, sink: PigeonEventSink<IndoorState>) {
        self.sink = sink
        sink.success(lastState)
    }

    override func onCancel(withArguments _: Any?) {
        sink = nil
    }
}

final class IndoorController: NSObject, _IndoorMessenger {
    private let mapboxMap: MapboxMap
    private let stream = IndoorEventStream()
    private var cancellable: AnyCancellable?

    init(withMapboxMap mapboxMap: MapboxMap, messenger: SuffixBinaryMessenger) {
        self.mapboxMap = mapboxMap
        super.init()

        IndoorUpdatesStreamHandler.register(with: messenger.messenger, instanceName: messenger.suffix, streamHandler: stream)

        cancellable = mapboxMap.indoor.onIndoorUpdated.sink { [weak self] state in
            guard let self else { return }
            let data = state.toFLT()
            stream.lastState = data
            stream.sink?.success(data)
        }
    }

    func selectFloor(floorId: String?) throws {
        mapboxMap.indoor.selectFloor(selectedFloorId: floorId)
    }

    func tearDown() {
        cancellable?.cancel()
        cancellable = nil
        stream.sink = nil
    }
}

private extension MapboxMaps.IndoorState {
    func toFLT() -> IndoorState {
        IndoorState(
            floors: floors.map { IndoorFloor(id: $0.id, name: $0.name) },
            selectedFloorId: selectedFloorId.isEmpty ? nil : selectedFloorId
        )
    }
}
